packages/babel-parser/src/parser/expression.js

```diff
@@ -2611,4 +2634,36 @@ export default class ExpressionParser extends LValParser {
 
     return ret;
   }
+
+  // https://github.com/tc39/proposal-js-module-blocks
+  parseModuleExpression(): N.ModuleExpression {
+    this.expectPlugin("moduleBlocks");
+    const node = this.startNode<N.ModuleExpression>();
+    this.next(); // eat "module"
+    const oldLabels = this.state.labels;
+    this.state.labels = [];
+    const oldExportedIdentifiers = this.state.exportedIdentifiers;
+    this.state.exportedIdentifiers = [];
+    this.eat(tt.braceL);
+    this.scope.enter(SCOPE_PROGRAM);
+    const oldUndefinedExports = this.scope.undefinedExports;
+    this.scope.undefinedExports = new Map();
+    let paramFlags = PARAM;
+    if (this.hasPlugin("topLevelAwait")) {
+      paramFlags |= PARAM_AWAIT;
+    }
+    this.prodParam.enter(paramFlags);
+    const oldInModule = this.inModule;
+    this.inModule = true;
+    const program = this.startNode<N.Program>();
+    node.body = this.parseProgram(program, tt.braceR, "module");
```

Comment:
We should also construct new `classScope` like we did in

https://github.com/babel/babel/blob/10978bb65a4b4e8874ca8dd3054b8c31b5838b7f/packages/babel-parser/src/parser/index.js#L37

otherwise it will mess up with upper class definition:

```js
class B {
  #p() {
    module {
      class C { [this.#p]; } // it should not have access to `#p` defined across the module block
    }
  }
}
```

Reply:
I wonder if it would be better to introduce a new `initializeScopes` function, so htat if we have to initialize more things we only have to update one place.

Maybe something like
```js
const oldScopes = this.initializeScopes();
node.body = this.parseProgram(program, tt.braceR, "module"));
Object.assign(this, oldScopes);
```
and then we re-use `initializeScopes` in `Parser`'s `constructor`.